runtime/include/QuantumDevice.hpp

```diff
@@ -241,8 +241,8 @@ struct QuantumDevice {
      * @param probs The pre-allocated `DataView<double, 1>`
      * @param wires Wires will restrict probabilities to a subset of the full system
      */
-    virtual void PartialProbs(DataView<double, 1> &probs,
-                              const std::vector<QubitIdType> &wires) = 0;
+    virtual void PartialProbs(DataView<double, 1> &probs, const std::vector<QubitIdType> &wires,
+                              const bool use_device_shots = true) = 0;
```

Comment:
Just thinking that we should be careful changing the API, since we have other devices that implement it now. If we add the breaking change maybe we take the opportunity revisit why sample and counts pass in the shots but none of the other MPs do?

I can see two options:
- every MP has a shot argument
  - consistent signature across all MPs
  - can customize the shot number when called from measure (or other internal contexts)
  - no need for a rather arbitrary `use_device_shots` in the Probs function only
- no MP has a shot argument, device shots are queried if needed by an MP
  - consistent signature across all MPs
  - need a different way to alter shots in certain contexts, maybe temporarily overwrite device shots?
  - no need for a rather arbitrary `use_device_shots` in the Probs function only
 
(And a third option I suppose where we don't touch the interface and still do option 2.)

Reply:
If it works, I would go with option 2 since it seems to have the least impact.

Reply:
Agree with @dime10 points here!

`PartialProbs` and `Probs` currently use the device-shots provided at the device initialization in the runtime.
There are two methods to set and get device shots in C++. An implementation for Option 2 can be via adding C-API instructions for these methods in the runtime and use them in the compiler to overwrite the device shots when is required throughout the lifetime of a device.

Alternatively, we can update the signature of `Probs` and `PartialProbs` to take `size_t shots` similar to `Counts`, `PartialCounts`, `Sample` and `PartialSample` in favour of making our MPs consistent. This is closer to the current approach but without using the arbitrary `use_device_shots` argument. I like this approach better here as it doesn't add new instructions to the C-API and it makes the signature of tensor-based MPs more consistent.

Reply:
Given that the need to alter the device shots only happens within the device in the first place, I don't think that requires any new c-api functions

Reply:
yeah, if this is the case that is only needed with the device initialization :+1:

Reply:
The problem was much simpler than what is proposed originally :smile: Looks good!